ts: Add IDL seed types

Shiva953 opened the pull request:
🤖 This pull request was created by https://app.gib.work/r/Ql3a0gKk/anchor 
 ⚡️ Payment for bounty is immediately sent to the contributor after merge.

Shiva953 committed:
added IdlSeed Type for IDL PDA seeds

Signed-off-by: Shiva953 <b22070@students.iitmandi.ac.in>

acheroncrypto commented on ts/packages/anchor/src/idl.ts:
```diff
@@ -67,7 +67,16 @@ export type IdlPda = {
 };
 
-export type IdlSeed = any; // TODO
+// Based on the rust type defined in lang/sync/src/idl/types.rs and the definitions in lang/syn/src/idl/pda.rs
+export type IdlSeed =
+  | { kind: 'Const'; value: IdlConstant }
+  | { kind: 'Arg'; value: IdlSeedArg }
+  | { kind: 'Account'; value: IdlAccount };
+
+export interface IdlSeedArg {
+  type: IdlType;
+  path: string;
+}
```
Hey, thanks for your contribution but I don't think these types are correct based on the Rust definitions you've linked as a comment.

Shiva953 replied:
Thanks for the feedback. Updated the types, defined separate ones for `IdlSeedConst`, `IdlSeedArg` and `IdlSeedAccount` based on the rust definitions. Let me know the changes required.

Shiva953 committed:
updated types

acheroncrypto commented on ts/packages/anchor/src/idl.ts:
```diff
@@ -67,7 +67,38 @@ export type IdlPda = {
+  value: SerdeJsonValue;
+};
+
+export type SerdeJsonValue =
+  | null
+  | boolean
+  | number
+  | string
+  | SerdeJsonValue[]
+  | { [key: string]: SerdeJsonValue };
```
Thanks for adding this but I think we can just use `any` here and remove this type.

Shiva953 replied:
sure!

acheroncrypto commented on ts/packages/anchor/src/idl.ts:
```diff
@@ -67,7 +67,38 @@ export type IdlPda = {
+export type IdlSeedAccount = {
+  kind: "account";
+  type: IdlType;
+  account?: string; // Adjust the type based on the actual type of account in your use case
```
```suggestion
  account?: string;
```

acheroncrypto commented on ts/packages/anchor/src/idl.ts:
```diff
@@ -67,7 +67,38 @@ export type IdlPda = {
 };
 
-export type IdlSeed = any; // TODO
+// Based on the rust type defined in lang/sync/src/idl/types.rs and the definitions in lang/syn/src/idl/pda.rs
```
```suggestion
```

acheroncrypto reviewed (commented):
Thanks, looks great!

Before we merge, could you also add a [CHANGELOG](https://github.com/coral-xyz/anchor/blob/master/CHANGELOG.md#features) entry for this?

Shiva953 committed:
removed the SerdeJsonValue type and added any

Shiva953 committed:
made required changes

Signed-off-by: Shiva953 <b22070@students.iitmandi.ac.in>

Shiva953 committed:
updated changelog

Signed-off-by: Shiva953 <b22070@students.iitmandi.ac.in>

Shiva953 commented:
@acheroncrypto any updates?

acheroncrypto commented:
TS package has build errors: https://github.com/coral-xyz/anchor/actions/runs/7368700143/job/20074262110?pr=2752

Shiva953 committed:
fixed build errors

Signed-off-by: Shiva953 <b22070@students.iitmandi.ac.in>

Shiva953 commented:
The tests now pass, added the necessary missing types in `IdlSeedConst` and `IdlSeedArg`, also modified the error message a bit(since `accountDesc.pda.programId.kind` included inside the `throw New Error()` was resulting in type `never` for `accountDesc.pda.programId`)

<img width="772" alt="Screenshot 2024-01-02 at 7 16 34 PM" src="https://github.com/coral-xyz/anchor/assets/120790871/86844d53-29f6-42a7-aada-81e1842f2b18">

Shiva953 committed:
removed console.log check

Signed-off-by: Shiva953 <b22070@students.iitmandi.ac.in>

acheroncrypto reviewed (changes requested):
There is also a [formatting error](https://github.com/coral-xyz/anchor/actions/runs/7386584794/job/20106102432?pr=2752), you can fix it by running `yarn lint:fix`.

Shiva953 committed:
fixed linting errors and changelog update

Signed-off-by: Shiva953 <b22070@students.iitmandi.ac.in>

Shiva953 committed:
Merge branch 'master' into gib-bounty-34HfM

acheroncrypto committed:
Add correct types

acheroncrypto committed:
Add back specifying the seed that caused the error

acheroncrypto reviewed (approved):
Thank you for the contribution!
